### ISSUE TITLE
Add Password reset

### ISSUE THREAD

**Issue by CarlosNZ** [openmsupply/application-manager-web-app#966]
Will require 2 simple templates, and a small codebase change:

### 1st template:

- enter username* - check its a valid username in system (use isUnique endpoint)
- onSubmit -- verification link sent to email of username
- On verification:

~~### 2nd template~~ (went with 1-template solution as per discussion below and explanation in [PR](#975))
~~- enter new password field~~
~~- onSubmit -- change password~~

### Changes to codebase:
- add a "forgot password" link on login screen that links directly to first template application (above)
---

\* would be good to allow reset by email address, but we're not enforcing unique email addresses currently, so wouldn't be able to tell which account it is for.

**Comment by andreievg** [comment]
> * would be good to allow reset by email address, but we're not enforcing unique email addresses currently, so wouldn't be able to tell which account it is for.

I think we should be

> check its a valid username in system 

Ideally action will do that ?

> 2nd template

Hmm, not really sure how we can do this, there a few options, and probably more options beyond this list:
* Add a very very long serial to application, send link directly to application (how do we restrict visibility of those applications, can't do it by session id ?) 
* Link also has sessionId=X (session id is sent to server when asking for userInfo, and is set as session ID for the user, application has that sessionID in the policy)

**Comment by CarlosNZ** [comment]
> > * would be good to allow reset by email address, but we're not enforcing unique email addresses currently, so wouldn't be able to tell which account it is for.
> 
> I think we should be

I dunno, I'm in two minds about this. Yes, it makes sense in most real-world situations, but when we're developing and doing demos, etc., it's a massive hassle to have to use a different email for every dummy account we create. (and also, most of our dummy accounts currently have no email address -- which is what we need, otherwise someone's going to get a lot of spam during development of templates)

> > check its a valid username in system
> 
> Ideally action will do that ?

Needs to do it live in the form, before submission -- and tell user if it can't find that username.

> > 2nd template
> 
> Hmm, not really sure how we can do this, there a few options, and probably more options beyond this list:

Yeah, I think something like your second option is probably the way to go -- I'll have a think and comment back with a more detailed suggestion.

**Comment by CarlosNZ** [comment]
@andreievg How about this:

- On submission of first template, create verification, as per above, and we save the username in the "data" field of the verification record.
- on verification (i.e. clicking link in email), the verification page has a link to start the new "change password" template -- which is created using the verification unique ID as its serial number (which automatically gives it a hard-to-find url)
- then in that application, we have a "is verified" check, based on looking up the serial/uniqueID in verification table, so if anyone manually starts an application of this type through other means, it won't be submittable. 
- the username for which the password is being changed can be passed through using the "data" field of the verification table, so no need for the user to be logged in to identify them.

So, basically the "change password" application needs two things in order to be submittable:
- username from verification record
- the "is_verified" field from that same verification record must be true (which means user has confirmed their account via email)

There might be some additional stumbling blocks, but I think I can test this approach fairly quickly and see if it works using our current infrastructure.

**Comment by andreievg** [comment]
> Needs to do it live in the form, before submission -- and tell user if it can't find that username.

I guess it's ok for now (typically, secure password resets are pretty 'quiet' about that sort of thing).

As for the suggested solution:

I think it would work, I think it's ok to allow submission, without verification (back end would need to do verification either way, including checking for is_verified and getting the user from verified record).

Btw how is the serial number going to match verification unique ID ? (unless it's created on back end, and link to new application is send in verification)

**Comment by CarlosNZ** [comment]
> I was thinking that it could be passed to "generateText" as part of "outputCumulative" (the "data" object on veriifcation), but now that I think about it, a newly created application has access to very little other than the current user data (which is non registered in this case). I've got a couple of ideas, but I'll need to test them out to see if I'm barking up the wrong tree or not 😄

Actually, might be easier to do with a single application -- that you submit twice.

The "password change" field is hidden until the verification is verified. So you submit with just a username, the status changes to "ChangesRequested".

The verification confirmation has a link back to the same application, but now there is a "password change" field visible due to the verification being confirmed, and you submit it again. This time it changes the password.

**Comment by andreievg** [comment]
> Actually, might be easier to do with a single application -- that you submit twice.

> The "password change" field is hidden until the verification is verified. So you submit with just a username, the status changes to "ChangesRequested".

> The verification confirmation has a link back to the same application, but now there is a "password change" field visible due to the verification being confirmed, and you submit it again. This time it changes the password.

I like this as a workflow, but my worry is, how do you prevent the following:
* Non registered user, with graphQL, looking up applications that are in between verification and submission (and changing passwords for other users)
* Bypassing, again with graphQL queries, the whole process and submitting a new password for another user ?

I think we need to associate a session with this application, and only resume it with a special link, i can't think of a way to do it otherwise. 

Just FYI the only other two places where user details/password is modified:
* User Registration -> this will only happen on verification link (it's possible to submit a raw graphQL with a different username, that already exists, but modify record action only inserts, since there is no id or matchField, and username is unique, so we are safe)
* User Edit -> It's possible to submit raw graphQL for another user and change their password/username (by modifying userId on application) <- this can be restricted in row level update policy (user id can only be jwt token user id)

**Comment by nmadruga** [comment]
Just following up with this now:

> Actually, might be easier to do with a single application -- that you submit twice.
> 
> The "password change" field is hidden until the verification is verified. So you submit with just a username, the status changes to "ChangesRequested".
> 
> The verification confirmation has a link back to the same application, but now there is a "password change" field visible due to the verification being confirmed, and you submit it again. This time it changes the password.

That seems to be a good solution - I also prefer this rather than the 2 templates solution. Although I think it might be tricky to do this change on statuses - usually we only allow for a Draft application to be modified so, when do we change these statuses might be something to get cleared.

> I like this as a workflow, but my worry is, how do you prevent the following:
> 
> * Non registered user, with graphQL, looking up applications that are in between verification and submission (and changing passwords for other users)
> * Bypassing, again with graphQL queries, the whole process and submitting a new password for another user ?
> 
> I think we need to associate a session with this application, and only resume it with a special link, i can't think of a way to do it otherwise.

Trying to see what the benefits are of using the sessionId in this process. So we can verify the same person that made the request for password change is the owner of the email entered. Right? Would this be ok if the link is opened in another tab, that would keep the same session? And do we already use that for the email confirmation or not?
